CanHazFunny/CanHazFunny.Tests/JesterTests.cs

```diff
@@ -1,7 +1,290 @@
+        // Arrange: Pass null to MockableJokeService to simulate GetString() returning null
+        var jokeService = new MockableJokeService(nextJoke: null);
+
+        // Act
+        var result = jokeService.GetJoke();
+
+        // Assert
+        Assert.Equal("No joke found", result);
+    }
+
+    [Fact]
+    public void GetJoke_ReturnsErrorMessage_OnError()
+    {
+        // Arrange: Covers the catch block
+        var jokeService = new MockableJokeService(shouldThrow: true);
+
+        // Act
+        var result = jokeService.GetJoke();
+
+        // Assert
+        Assert.Equal("Error retrieving joke. Please try again later", result);
+    }
+}
+
+// ----------------------------------------------------------------------
+// CONSOLE OUTPUT TESTS
+// ----------------------------------------------------------------------
+
+public class ConsoleOutputTests
+{
+    [Fact]
+    public void WriteLine_FormatsMessageCorrectly()
+    {
+        // Arrange
+        var outputService = new ConsoleOutput();
+        var testMessage = "This is a test joke";
+        var stringWriter = new StringWriter();
+        Console.SetOut(stringWriter);
+
+
+        // Act
+        ((IOutput)outputService).WriteLine(testMessage);
+
+
+        // Assert
+        var output = stringWriter.ToString().Trim();
+        Assert.Equal(testMessage, output);
+    }
+
+    [Fact]
+    public void WriteLine_StoresOutputInPublicProperty()
+    {
+        // Arrange
+        var outputService = new ConsoleOutput();
+        var testMessage = "Another test";
+        var stringWriter = new StringWriter();
+        Console.SetOut(stringWriter);
+
+
+        // Act
+        ((IOutput)outputService).WriteLine(testMessage);
+
+
+        // Assert
+        Assert.Equal(testMessage, outputService.Output);
+    }
+}
+
+// ----------------------------------------------------------------------
+// INTEGRATION TESTS (Execute)
+// ----------------------------------------------------------------------
+
+public class IntegrationTests
+{
+    [Fact]
+    public void JokeServiceImplementation_GetJoke_CallsUnderlyingJokeService()
+    {
+        // This test makes a real network call
+        var jokeServiceImpl = new JokeService();
+
+
+        // Act
+        var joke = jokeServiceImpl.GetJoke();
+
+
+        // Assert
+        Assert.NotNull(joke);
+        Assert.NotEmpty(joke);
+    }
+
+    [Fact]
+    public void Jester_IntegrationTest_WithRealServices()
+    {
+        // This test makes a real network call
+        var stringWriter = new StringWriter();
+        Console.SetOut(stringWriter);
+        var outputService = new ConsoleOutput();
+        var jokeService = new JokeService(); 
+        var jester = new Jester(outputService, jokeService);
+
+
+        // Act
+        jester.TellJoke();
+
+
+        // Assert
+        var consoleOutput = stringWriter.ToString().Trim();
+        Assert.NotEmpty(outputService.Output!);
```

Comment:
Could combine with the other change I suggested for consoleOutput.
```suggestion
        Assert.NotEmpty(consoleOutput);
```
The code I suggested potentially might not work out of the box. If that's the case, perhaps a Mock interface could catch the string for testing.
Let me know if I’m missing anything.